tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1178,186 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
+    RESULT_ENSURE_EQ(*negotiated_cipher, NULL);
+
+    const struct s2n_security_policy *security_policy;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &security_policy));
+    const struct s2n_cipher_preferences* cipher_preferences = security_policy->cipher_preferences;
+    RESULT_ENSURE_REF(cipher_preferences);
+
+    bool in_group = false;
+    /** 
+     * Tracks mutually supported cipher suite in the client index. This is needed to track tie-breakers
+     * in an equal-preference group.
+     */
+    int negotiated_client_index = count;
+    /** 
+     * The highest protocol version match cipher. This is a fall back cipher if
+     * no other mutually-supported cipher suite is identified.
+     */
+    struct s2n_cipher_suite* highest_vers_match_cipher = NULL;
+
+    for (uint32_t i = 0; i < cipher_preferences->count; i++) {
+        struct s2n_cipher_suite* server_cipher = cipher_preferences->suites[i];
+        RESULT_ENSURE_REF(server_cipher);
+
+        /* Check if the cipher suite is an equal-preference grouping delimiter */
+        if (server_cipher == &s2n_equal_preference_group_start) {
+            in_group = true;
+            continue;
+        } 
+        if (server_cipher == &s2n_equal_preference_group_end) {
+            in_group = false;
+            /* Exiting a group and a negotiated cipher has already been found. */
+            if (*negotiated_cipher) {
+                return S2N_RESULT_OK;
+            }
+            continue;
+        }
+
+        /* If connection is for SSLv3, use SSLv3 version of suites */
+        if (conn->client_protocol_version == S2N_SSLv3) {
+            server_cipher = server_cipher->sslv3_cipher_suite;
+        }
+
+        /* Cipher suite is NOT a delimiter */
+        int client_index = s2n_wire_index_of_cipher_suite(server_cipher->iana_value, wire, count, cipher_suite_len);
+
+        /* Client does not support this cipher. Skip. */
+        if (client_index < 0) {
+            continue;
+        }
+
+        /* Validate that the cipher suite is suitable for this connection. */
+        if (!s2n_cipher_suite_match_is_valid(conn, server_cipher)) {
+            continue;
+        }
+
+        /* Don't immediately choose a cipher the connection shouldn't be able to support */
+        if (conn->actual_protocol_version < server_cipher->minimum_required_tls_version) {
+            if (!highest_vers_match_cipher) {
+                highest_vers_match_cipher = server_cipher;
+            }
+            continue;
+        }
+        if (in_group) {
+            /* Store the most-preferred client cipher in-group */
+            if (client_index < negotiated_client_index) {
+                negotiated_client_index = client_index;
+                *negotiated_cipher = server_cipher;
+            }
+        } else {
+            /* Else, the client and server both support a non-grouped cipher. */
+            *negotiated_cipher = server_cipher;
+            break;
+        }
```

Comment:
We try to avoid nesting conditionals / loops. The deeper you go, the more difficult it is to reason about the code, particularly visually.

```suggestion
        if (in_group && client_index < negotiated_client_index) {
           negotiated_client_index = client_index;
           *negotiated_cipher = server_cipher;
        } else {
            /* Else, the client and server both support a non-grouped cipher. */
            *negotiated_cipher = server_cipher;
            break;
        }
```

Alternatively, to highlight that "in_group" is the exception instead of the norm:
```suggestion
        if (in_group && client_index < negotiated_client_index) {
           negotiated_client_index = client_index;
           *negotiated_cipher = server_cipher;
           continue;
        } 
        
        *negotiated_cipher = server_cipher;
```

Reply:
Isn't your new logic different though?

Reply:
That change would break the current logic. The current if else first checks 'if I'm in a group, else do something different'. Your change would be 'if I'm in a group and I found a suitable client index, else do something'.

Reply:
Yeah sorry, github suggestions are not the best place for refactoring. But I'm sure you can figure something out that at least doesn't have the standard logic in an `else`. Also avoid `break`: this loop is really too long to be mixing 'continue' and 'break' without someone messing it up.

Another attempt at github suggestion refactoring:
```suggestion
        if (in_group)
           if (client_index < negotiated_client_index) {
              negotiated_client_index = client_index;
              *negotiated_cipher = server_cipher;
           }
           continue;
        } 
        
        *negotiated_cipher = server_cipher;
```